server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java

```diff
@@ -63,6 +67,16 @@ public BucketsAggregator(String name, AggregatorFactories factories, SearchConte
             multiBucketConsumer = (count) -> {};
         }
         docCounts = bigArrays.newIntArray(1, true);
+
+        // Check index mappings to find a field of type doc_count. If one is found
+        // use that one to retrieve the doc count for the bucket
+        for (MappedFieldType fieldType : context.getQueryShardContext().getMapperService().fieldTypes()) {
+            if (DocCountFieldMapper.CONTENT_TYPE.equals(fieldType.typeName())) {
+                // If a field of type doc_count has been found, use it to provide the bucket doc_count values
+                fieldDocCountProvider = new FieldBasedDocCountProvider(fieldType.name());
```

Comment:
Using a canonical name for the Lucene field could simplify this part ? We could use the `FieldBasedDocCountProvider` all the time and the logic to handle missing values/field would be implemented directly in this concrete class (no need to introduce an interface) ?

Reply:
With the interface, I tried to have a more generic approach. Since we are moving to the concept of a summary "document" maybe in the future we will need more implementations. For example, extract `doc_count` from a script. Another case would be to use the `histogram` buckets count as the `doc_count`. 

Also, in 99% of the cases the good old implementation of incrementing by 1 is what we need. In this  case, reading fields and handling missing values would be more overhead. Resolving the correct `DocCountProvider` instance once in the `SearchContext` is minimal overhead imho.

Reply:
> Also, in 99% of the cases the good old implementation of incrementing by 1 is what we need. In this case, reading fields and handling missing values would be more overhead. 

We'd just need to check if the `_doc_count` field exists in the Lucene segment ? If not, then we can fallback to return 1 all the  time. My point here is that it would simplify things to expect a canonical field name in Lucene.

Reply:
I pushed commit 39c43a0. 

`DocCountFieldMapper` uses canonical name `_doc_count` for the Lucene field and I modified `FieldBasedDocCountProvider` to only look for this field. It already had the logic to handle missing values/field in the Lucene segment. 

As suggested, I modified `BucketAggregator` to use `FieldBasedDocCountProvider`  by default, without looking up the index mapping for a `doc_count` field. 

So, I have the following questions: 

1. Is there a significant performance overhead for looking up a field for each lucene segment? Especially, when 99% of the times the `_doc_count` field is missing?

2. If performance overhead is negligible, does it still make sense to instantiate `FieldBasedDocCountProvider` in the `SearchContext`? I hesitate leaking this implementation aspect of the `BucketAggregator` to the `SearchContext`.

Reply:
> 1

I think it's ok and will be sometimes better than checking every field in the mapping.

> 2

I agree that it's easier to create a new iterator for each multi-bucket aggregator. Maybe that's the best approach to start with since accessing these values should be fast. Although, if we keep the current design, I think we can use a plain `NumericDocValues` rather than introducing the `DocCountProvider` interface.

Reply:
> I think it's ok and will be sometimes better than checking every field in the mapping.

Agreed. I will go down that road. Although the initial implementation would check mapping once per shard and not once per segment. 

> Although, if we keep the current design, I think we can use a plain NumericDocValues rather than introducing the DocCountProvider interface.

For the sake of the separation of concerns principle, I would keep the `FieldBasedDocCountProvider` instead of using bare bones `NumericDocValues`. I think it makes code cleaner and more testable. Also, this is a class we can pull up at `SearchContext` if we want at a later stage. If you feel that there's too much abstraction, I can remove the `DocCountProvider` interface and only keep the concrete class.